### PR TITLE
Reconstruct Map from URL

### LINKED ISSUE #1333: Re-construct Map with a Formatted Shareable URL
Given a shareable URL of map previously distorted and knitted using the Mapknitter lite service, users of the service have no functionality to re-construct the map and restore it to its previous state on the tile layer. This limits user experience and removes a level of convenience users have now gotten used to having used other web services. This issue builds on issue #1330 and PR #1334.

**Please describe the desired behavior.**
Users should have the flexibility of re-constructing map into their previous state by loading shareable URL in a browser.

### PR DESCRIPTION
**Pull request by segun-codes**
Fixes #1333 (<=== Add issue number here)

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [X] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `grunt test`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updates
* [x] @mention the original creator of the issue in a comment below for help or for a review

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

### CODE REVIEW

**Comment by segun-codes** [comment]
@jywarren, just for you to see progress made so far and probably offer some advice, work not fully ready yet.

File: src/DistortableCollection.js [hunk 1]
```diff
@@ -199,27 +201,109 @@ L.DistortableCollection = L.FeatureGroup.extend({
...
+  // OPTION 2 - ASYNC/AWAIT
+  readJSON(currentURL) {
+    let jsonDownloadURL;
+
+    if(this._isJsonDetected(currentURL)) {
+      jsonDownloadURL = this._extractJSONDownloadURL(currentURL);
+    }
+
+    if (jsonDownloadURL) {
+      this._getImageCollectionProps(jsonDownloadURL)
```
Comment (jywarren) [inline src/DistortableCollection.js:290]: await

File: examples/js/shareable.js [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+form.addEventListener('submit', (event) => {
+  event.preventDefault();
+  extractKey();
+});
+
+// delete, it's for testing purposes only
+generateExportBtn.addEventListener('click', (e) => {
+  console.log('generateExportBtn clicked');
+  const url = 'http://localhost:8081/examples/archive.html?json=https://archive.org/download/segetest/segetest.json';
+  console.log('map.imgGroup.readJSON(url): ', map.imgGroup.readJSON(url));
```
Comment (jywarren) [inline examples/js/shareable.js:47]: await

File: examples/js/shareable.js [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+L.Control.geocoder().addTo(map);
+setupCollection();
+
+form.addEventListener('submit', (event) => {
+  event.preventDefault();
+  extractKey();
+});
+
+// delete, it's for testing purposes only
+generateExportBtn.addEventListener('click', (e) => {
```
Comment (jywarren) [inline examples/js/shareable.js:44]: async

Comment (segun-codes) [reply examples/js/shareable.js:44]: Thanks.

File: examples/js/shareable.js (hunk 2 above)
Comment (segun-codes) [reply examples/js/shareable.js:47]: Thanks!

File: src/DistortableCollection.js (hunk 1 above)
Comment (segun-codes) [reply src/DistortableCollection.js:290]: Yeah, many thanks!

**Comment by segun-codes** [comment]
@jywarren, this PR is now ready for your review. You may refer to illustrations 1 & 2 below and then offer your thoughts. Many thanks!

**Illustration 1:**
URLs used for testing:
1. http://localhost:8081/examples/shareable.html?json=https://archive.org/download/segeotest/segeotest.json
2. http://localhost:8081/examples/shareable.html?json=https://archive.org/download/segeotest2/segeotest2.json
3. http://localhost:8081/examples/shareable.html?json=https://archive.org/download/segeotest3/segeotest3.json
4. http://localhost:8081/examples/shareable.html?json=https://archive.org/download/segeotest4/segeotest4.json

**Illustration 2:**
Illustration

File: src/DistortableCollection.js [hunk 4]
```diff
@@ -199,19 +199,65 @@ L.DistortableCollection = L.FeatureGroup.extend({
     return reduce / imgs.length;
   },
 
-  isJsonDetected(currentURL) {
+  _extractJSONDownloadURL(currentURL) {
```
Comment (jywarren) [inline src/DistortableCollection.js:202]: Here I'm thinking... maybe this belongs outside the core lib, since it addresses the URL bar?

File: src/DistortableCollection.js [hunk 5]
```diff
@@ -199,19 +199,65 @@ L.DistortableCollection = L.FeatureGroup.extend({
...
-        console.log('JSON found in map shareable link');
+        console.log('JSON found in map shareable link'); // left here purposely
         return true;
       }
     }
+
     return false;
   },
 
+  async _generateFromJson(jsonDownloadURL) {
```
Comment (jywarren) [inline src/DistortableCollection.js:223]:
```suggestion
  async _recreateMapFromJson(json) {
```

How about this for a name?

And, what do you think of separating the remote fetching from the regenerating, because we could then use the regeneration code on JSON from another source, like localStorage? So we could have:

`_recreateMapFromJson(json)`

and

`_recreateMapFromJsonUrl(url)` which just fetches the URL and runs `_recreateMapFromJson(json)` on the response?

File: src/DistortableCollection.js [hunk 6]
```diff
@@ -199,19 +199,65 @@ L.DistortableCollection = L.FeatureGroup.extend({
...
+      imgCollectionProps[index] = response.data.images;
+
+      return {avg_cm_per_pixel: response.data.avg_cm_per_pixel, imgCollectionProps};
+    } catch (err) {
+      console.log('err', err);
+    }
+  },
+
+  // expects URL in this format: 'http://localhost:8081/examples/shareable.html?json=https://archive.org/download/segeotest/segeotest.json'
+  async generateFromJsonUrl(currentURL) {
```
Comment (jywarren) [inline src/DistortableCollection.js:245]: Here, I think we should be doing this in the examples folder, since another application using this library might already be using the URL for its purposes.

**Review by jywarren — CHANGES_REQUESTED** [review]
Hi @segun-codes this is exciting! It's looking really good. I made some suggestions, first to separate out fetching from recreating the map, and second, to move all URL-related code outside the core library. What do you think?

**Comment by segun-codes** [comment]
> Hi @segun-codes this is exciting! It's looking really good. I made some suggestions, first to separate out fetching from recreating the map, and second, to move all URL-related code outside the core library. What do you think?

Okay, so this is also a good idea. 

You know I figure if we agree that the sharing feature (generating JSON and reconstructing maps from the JSON) is core, then it should be natively supported and shipped in the library. This feature is so basic nowadays that the library users will naturally expect to expend quite minimal effort to wire it up using proper documentation. Externalizing the codes into examples folder means more work for users but then we can still follow this route if it's relatively more beneficial.

File: src/DistortableCollection.js (hunk 4 above)
Comment (segun-codes) [reply src/DistortableCollection.js:202]: Okay, but you know it depends on what we agree should be core to the library and what responsibility should be passed to the library user.

File: src/DistortableCollection.js (hunk 5 above)
Comment (segun-codes) [reply src/DistortableCollection.js:223]: Okay, sounds good. Thanks!

File: src/DistortableCollection.js (hunk 6 above)
Comment (segun-codes) [reply src/DistortableCollection.js]: Okay, thanks.

**Comment by segun-codes** [comment]
Hi @jywarren, the code has been restructured (and some renaming done) as per your advice, Could you kindly take a look, many thanks!

**Comment by jywarren** [comment]
Hi @segun-codes --

> You know I figure if we agree that the sharing feature (generating JSON and reconstructing maps from the JSON) is core, then it should be natively supported and shipped in the library. This feature is so basic nowadays that the library users will naturally expect to expend quite minimal effort to wire it up using proper documentation. Externalizing the codes into examples folder means more work for users but then we can still follow this route if it's relatively more beneficial.

I hear your point for sure. However, I want to suggest that Leaflet.DistortableImage is designed as a library, not an application. This is the slightly odd thing about MapKnitter Lite, that it's an application we are developing inside the library repository. We could at some point decide that we want a separate, stand-alone application repository, like https://github.com/publiclab/mapknitter.js, or https://github.com/publiclab/mapknitter-lite. But for now, we have to draw those distinctions ourselves conceptually and in where we place the code. 

For MapKnitter Lite it definitely makes sense to include the URL code. But for other applications using the Leaflet.DistortableImage (LDI) library, they may have their own logic, or their own parameters or routing. Many web applications use URL parameters, and they may do so in different ways. I think it's reasonable to ask LDI to read a remote JSON file. But to read and write the current URL could easily conflict with other applications' logic, and create incompatibilities. That's how I'm thinking about it, I hope it makes sense. I think we wouldn't be in this position if we'd decided earlier to clearly break out MK Lite into its own application. But if we keep our code well organized, we can choose to do that in the future. What do you think?

File: examples/js/shareable.js [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+
+    return {avg_cm_per_pixel: response.data.avg_cm_per_pixel, imgCollectionProps};
+  } catch (err) {
+    console.log('err', err);
+  }
+}
+
+// 1. Performs preliminary checks and obtains map data (needed to reconstruct map in another function)
+// expects URL in this format: http://localhost:8081/examples/shareable.html?json=https://archive.org/download/segeotest/segeotest.json
+async function recreateMapDataFromJsonUrl(url) {
```
Comment (jywarren) [inline examples/js/shareable.js:232]: This function, I think, could go in the core library. We could move it in a separate PR, though if you prefer!

**Review by jywarren — CHANGES_REQUESTED** [review]
This looks super. Really nicely formatted and documented code. Thank you!

I think you had mentioned that shareable.js was a temporary solution, and I wonder if it's worth trying to re-integrate your new code with archive.js now. Otherwise we'll be creating a new file just to remove it again later. But it's your choice!

I did mark one function that I think could go in the core library, as it wouldn't interrupt other application logic. But again its your choice!

**Comment by segun-codes** [comment]
> Hi @segun-codes --
> 
> > You know I figure if we agree that the sharing feature (generating JSON and reconstructing maps from the JSON) is core, then it should be natively supported and shipped in the library. This feature is so basic nowadays that the library users will naturally expect to expend quite minimal effort to wire it up using proper documentation. Externalizing the codes into examples folder means more work for users but then we can still follow this route if it's relatively more beneficial.
> 
> I hear your point for sure. However, I want to suggest that Leaflet.DistortableImage is designed as a library, not an application. This is the slightly odd thing about MapKnitter Lite, that it's an application we are developing inside the library repository. We could at some point decide that we want a separate, stand-alone application repository, like https://github.com/publiclab/mapknitter.js, or https://github.com/publiclab/mapknitter-lite. But for now, we have to draw those distinctions ourselves conceptually and in where we place the code.
> 
> For MapKnitter Lite it definitely makes sense to include the URL code. But for other applications using the Leaflet.DistortableImage (LDI) library, they may have their own logic, or their own parameters or routing. Many web applications use URL parameters, and they may do so in different ways. I think it's reasonable to ask LDI to read a remote JSON file. But to read and write the current URL could easily conflict with other applications' logic, and create incompatibilities. That's how I'm thinking about it, I hope it makes sense. I think we wouldn't be in this position if we'd decided earlier to clearly break out MK Lite into its own application. But if we keep our code well organized, we can choose to do that in the future. What do you think?

Okay, I get your point and it's fine. Many thanks!

File: examples/js/shareable.js (hunk 7 above)
Comment (segun-codes) [reply examples/js/shareable.js:232]: Okay, so I will quickly fix this, no need for fresh PR. Many thanks!

**Comment by segun-codes** [comment]
> This looks super. Really nicely formatted and documented code. Thank you!
> 
> I think you had mentioned that shareable.js was a temporary solution, and I wonder if it's worth trying to re-integrate your new code with archive.js now. Otherwise we'll be creating a new file just to remove it again later. But it's your choice!
> 
> I did mark one function that I think could go in the core library, as it wouldn't interrupt other application logic. But again its your choice!

I am happy to fix this too. I will move everything to archive.js.

File: examples/js/shareable.js (hunk 7 above)
Comment (segun-codes) [reply examples/js/shareable.js:232]: Hi @jywarren, still awaiting your reaction to the question. Thanks!

Comment (jywarren) [reply examples/js/shareable.js]: Can we change `recreateMapDataFromJsonUrl()` so it accepts simply a direct URL to a JSON file, instead of `archive.html?json=https://example.com/map.json` format? Then we can move `recreateMapDataFromJsonUrl()` without moving `extractJSONDownloadURL`. 

Maybe `extractJSONDownloadURL` could be renamed `extractJsonFromUrlParams` to clarify, and `recreateMapDataFromJsonUrl` could be `recreateMapFromJsonUrl`? (just a little shorter and we are recreating the map itself, not just the data? Not a big deal but a good balance of descriptive and not extremely long :laugh:)

Comment (jywarren) [reply examples/js/shareable.js]: And fantastic work on this!!!

File: examples/js/archive.js [hunk 8]
```diff
@@ -164,10 +172,13 @@ restoreWelcomeModal.addEventListener('click', (event) => {
 
 mapToggle.addEventListener('click', (event) => {
   new bootstrap.Offcanvas(sidebar).show();
+  sidebarOpen = true;
 });
 
 tileMap.addEventListener('click', (event) => {
-  bootstrap.Offcanvas.getInstance(sidebar).hide();
+  if (sidebarOpen) {
+   bootstrap.Offcanvas.getInstance(sidebar).hide();
```
Comment (jywarren) [inline examples/js/archive.js:180]: Just looked this up and I do see Offcanvas has a "toggle" method? Could save a few lines of code maybe!

https://getbootstrap.com/docs/5.0/components/offcanvas/#methods

**Comment by jywarren** [comment]
This is going to be super! 

I wanted to ask @segun-codes and @7malikk to think on this, maybe on Monday - and we can discuss at our Tuesday meeting. How might we think about a follow-up PR, which could abstract/modularize a lot of this code into a separate file? What are some ideas? We could brainstorm in a new issue maybe?

**Comment by segun-codes** [comment]
> This is going to be super!
> 
> I wanted to ask @segun-codes and @7malikk to think on this, maybe on Monday - and we can discuss at our Tuesday meeting. How might we think about a follow-up PR, which could abstract/modularize a lot of this code into a separate file? What are some ideas? We could brainstorm in a new issue maybe?

Okay, that's fine.

File: examples/js/shareable.js (hunk 7 above)
Comment (segun-codes) [reply examples/js/shareable.js]: Okay, so meaning it's okay to move `isJsonDetected(url)` and `fetchRemoteJson(jsonDownloadURL)` along with `recreateMapDataFromJsonUrl()` right? Of course, the renaming will be done too.

File: examples/js/archive.js (hunk 8 above)
Comment (segun-codes) [reply examples/js/archive.js:180]: Okay...thanks!

File: examples/js/shareable.js (hunk 7 above)
Comment (jywarren) [reply examples/js/shareable.js]: If recreateMapFromJsonUrl accepts only a direct URL and not a json= parameter, we wouldn't need to move isJsonDetected into the core library, since the url parameters are all managed outside in the application code, isn't that right? But I could be mixed up. But yes, fetchRemoteJson could move inside the core library too. Thank you!

Comment (segun-codes) [reply examples/js/shareable.js]: Okay, so that's correct. Mnay thanks!

File: examples/js/archive.js (hunk 8 above)
Comment (segun-codes) [reply examples/js/archive.js:180]: The "toggle" method works but then it generates an unintended effect when applied on the tile layer event listener. Clicking on tile layer should only close the sidebar if open and nothing more but using the "toggle" method means, the sidebar is switched on and off whenever the tile layer receives a click event. For this reason, I think we don't need the toggle method.

File: examples/js/shareable.js (hunk 7 above)
Comment (segun-codes) [reply examples/js/shareable.js]: > Can we change `recreateMapDataFromJsonUrl()` so it accepts simply a direct URL to a JSON file, instead of `archive.html?json=https://example.com/map.json` format? Then we can move `recreateMapDataFromJsonUrl()` without moving `extractJSONDownloadURL`.
> 
> Maybe `extractJSONDownloadURL` could be renamed `extractJsonFromUrlParams` to clarify, and `recreateMapDataFromJsonUrl` could be `recreateMapFromJsonUrl`? (just a little shorter and we are recreating the map itself, not just the data? Not a big deal but a good balance of descriptive and not extremely long :laugh:)

Hi @jywarren, as per renaming `recreateMapDataFromJsonUrl` to `recreateMapFromJsonUrl`,  if you look at the function again, you will discover we are not re-creating map but reproducing data required to construct the map. Map is actually reconstructed in the user application (e.g., in our case archive.js) with the `createMap()`. This is where we create instance of the `distortableImageOverlay`. So it's my considered opinion that `recreateMapDataFromJsonUrl` (or the like) is a more appropriate name. *What do you think?*

Comment (jywarren) [reply examples/js/shareable.js]: I see, and it makes sense. You could say "overlay" or "images" instead of "MapData"? But either is fine! Thank you!!

Comment (segun-codes) [reply examples/js/shareable.js]: Okay, thanks.

**Comment by segun-codes** [comment]
Hi @jywarren, renaming effected. Do you want to consider for merging. Many thanks!

File: examples/js/archive.js [hunk 9]
```diff
@@ -178,16 +189,100 @@ function getImageName(imageURL) {
...
+    if (fileExtension === 'json') {
+      console.log('JSON found in map shareable link'); // left here purposely
+      return true;
+    } 
+  }
+
+  return false;
+}
+
+function createMap (imageURL, options, newImage = false) {
```
Comment (jywarren) [inline examples/js/archive.js:213]: Could this be called createImage? Or placeImage?

Comment (jywarren) [reply examples/js/archive.js:213]: And could it be within the core library? We aren't doing anything custom here, right? Or do you need it for other purposes in archive.js?

File: examples/js/archive.js [hunk 10]
```diff
@@ -178,16 +189,100 @@ function getImageName(imageURL) {
...
+              tooltipText: imageObj.tooltipText,
+              corners: imageObj.nodes,
+            };
+            createMap(imageURL, options, false);
+          });
+      
+          return;
+        }
+
+        const imageObj = imageCollectionObj.imgCollectionProps[0];
```
Comment (jywarren) [inline examples/js/archive.js:264]: I'm sorry, I'm not quite following this section, why do we need to have two sections where we create an overlay? Maybe it needs an explanatory comment? Thanks!

Also, could we move all of this into recreateImagesFromJsonUrl?

**Review by jywarren — COMMENTED** [review]
Almost there!!!

File: examples/js/archive.js (hunk 9 above)
Comment (segun-codes) [reply examples/js/archive.js:213]: > And could it be within the core library? We aren't doing anything custom here, right? Or do you need it for other purposes in archive.js?

It shouldn't be in the core library because what it does is instantiate `distortableImageOverlay`. This is a user application responsibility. As for renaming, I will do that.

File: examples/js/archive.js (hunk 10 above)
Comment (segun-codes) [reply examples/js/archive.js:264]: **Two sections for overlay:** first section is needed because of the need to iteratively create overlays in cases where multiple images are to be retrieved from json file. The second section comes in handy when only one overlay is to be retrieved from the json file.

**Move to recreateImagesFromJsonUrl:** Like you suggested earlier which I have since agreed to and refactored the code accordingly (i.e., to it current state), I think this is unnecessary as it's a user application responsibility. The library simply supplies the data required to reconstruct overlays to their previous state, the user then has the flexibility to place the overlays however he wants using available API.  What do you think?

File: examples/js/archive.js (hunk 9 above)
Comment (segun-codes) [reply examples/js/archive.js:213]: > > And could it be within the core library? We aren't doing anything custom here, right? Or do you need it for other purposes in archive.js?
> 
> It shouldn't be in the core library because what it does is instantiate `distortableImageOverlay`. This is a user application responsibility or what do you think?. As for renaming, I will do that.

File: examples/js/archive.js (hunk 10 above)
Comment (jywarren) [reply examples/js/archive.js:264]: Ok that reasoning sounds good to me! Let me know and I can hit merge as soon as you like. Thanks!!

Comment (segun-codes) [reply examples/js/archive.js:266]: Hi @jywarren, you may now consider for merging. Many thanks!

**Comment by jywarren** [comment]
Amazing!!! Congrats, fantastic work!

**Comment by segun-codes** [comment]
> Amazing!!! Congrats, fantastic work!

Many thanks!

**Comment by 7malikk** [comment]
Grat work on @segun-codes!! 🎉

**Comment by segun-codes** [comment]
> Grat work on @segun-codes!! 🎉

Many thanks!!!
